mmaction/core/evaluation/average_precision.py

```diff
@@ -0,0 +1,315 @@
+import os.path as osp
+import pickle
+from collections import defaultdict
+
+import numpy as np
+
+from .bbox_overlaps import iou2d, spatio_temporal_iou3d, spatio_temporal_nms3d
+
+
+def pr_to_ap(precision_recall):
+    """Compute AP given precision-recall.
+
+    Args:
+        precision_recall (np.ndarray): precision_recall is an Nx2 array with
+            first row being precision and second row being recall.
+
+    Returns:
+        np.ndarray: The result of average precision.
+    """
+
+    pr_diff = precision_recall[1:, 1] - precision_recall[:-1, 1]
+    pr_sum = precision_recall[1:, 0] + precision_recall[:-1, 0]
+
+    return np.sum(pr_diff * pr_sum * 0.5)
+
+
+def frame_ap(det_results, labels, videos, gt_tubes, threshold=0.5):
+    """Calculate frame mAP for tubes.
+
+    Args:
+        det_results (np.ndarray): Detection results for each frames.
+        labels (list): List of action labels.
+        videos (list): List of video names.
+        gt_tubes (dict): Ground truth tubes for each video.
+        threshold (float): Threshold for IoU. Default: 0.5.
+
+    Returns:
+        float: The calculated frame mAP.
+    """
+    results = []
+    for label_index, label in enumerate(labels):
+        det_result = det_results[det_results[:, 2] == label_index, :]
+
+        gt = defaultdict(list)
+        for video_id, video in enumerate(videos):
+            tubes = gt_tubes[video]
+
+            if label_index not in tubes:
+                continue
+
+            for tube in tubes[label_index]:
+                for i in range(len(tube)):
+                    key = (video_id, int(tube[i, 0]))
+                    gt[key].append(tube[i, 1:5].tolist())
```

Comment:
what does the columns of tube represent?

Reply:
tube is a np.ndarray with (N, 5) shape, each row contains frame index + a bbounding box.

Reply:
Maybe i can add it in docstring